server/src/handlers/upload.rs

```diff
@@ -50,7 +60,7 @@ pub async fn upload_handler(
 
         std::fs::create_dir_all(&appstate.config.uploads_path)?;
```

Comment:
I think most of the logic here should be moved to the `FileStore` impl. For example, we're calling `create_dir_all` even if the user is using S3, right? And we're creating the `fs_file_id`, whereas we might not need it for S3 uploads.

Reply:
That makes sense to me.

Reply:
In this case actually, we are creating an intermediary file on the file system. Typically, if I knew this was a unix host, I would just write the tmp file into `/tmp`, but since we don't know the OS of the host I feel like this makes the most sense at this time. It also provides a fallback in the sense that if a user uploads a file and the s3 upload fails, the file still exists on the server. If the s3 upload succeeds, the file is deleted from the file system. Of course, there are pros and cons to this.

Reply:
I believe I could instead write directly to s3 in the multipart upload handler instead of using the intermediary file. My original thought is this might be more brittle when working with large files vs the current implementation where we essentially "copy" the file from fs to s3. But I will also add that while I've worked with S3 a lot in production code, I have never just "proxied" s3 uploads, typically just used a signed upload URL sent to client or run different processing on a file before uploading to s3.  So my hesitation to directly upload to s3 could just be unfamiliarity on my part.